Sem. 03/Solutions/Task01.cpp

```diff
@@ -0,0 +1,200 @@
+#include <iostream>
+#include <fstream>
+
+const char* UNSORTED_FILE_NAME = "unsortedDeliveries.bin";
+const char* SORTED_FILE_NAME = "sortedDeliveries.bin";
+const char* ERROR_MESSAGE = "Error";
+
+const int ADDRESS_MAX_LENGTH = 25;
+
+struct Delivery {
+    double price;
+    char address[ADDRESS_MAX_LENGTH];
+};
+
+int stringLength(const char* str) {
+    int length = 0;
+
+    for (int i = 0; str[i] != '\0'; ++i) {
+        length++;
+    }
+
+    return length;
+}
+
+int minStringLength(const char* str1, const char* str2) {
+
+    if(stringLength(str1) < stringLength(str2)) {
+        return stringLength(str1);
+    } else {
+        return stringLength(str2);
+    }
+}
+
+bool isSecondStringLarger(const char* str1, const char* str2) {
+    int minLength = minStringLength(str1, str2);
+
+    for (int i = 0; i < minLength; ++i) {
+        if (str1[i] == str2[i]) {
+            continue;
+        }
+
+        return str1[i] < str2[i];
+    }
+
+    if(stringLength(str2) > minLength) {
+        return true;
+    }
+
+    return false;
+}
+
+size_t getFileSize(std::ifstream& in) {
+
+    size_t currentPosition = in.tellg();
+    in.seekg(0, std::ios::end);
+    size_t fileSize = in.tellg();
+    in.seekg(currentPosition, std::ios::beg);
+
+    return fileSize;
+}
+
+namespace deliveryFunctions {
+
+    void assignAddress(Delivery& delivery, const char* address) {
+        int length = stringLength(address) + 1;
+
+        for (int i = 0; i < length; ++i) {
+            delivery.address[i] = address[i];
+        }
+    }
+
+    void swapDeliveries(Delivery& delivery1, Delivery& delivery2) {
+        Delivery temp = delivery1;
+        delivery1 = delivery2;
+        delivery2 = temp;
+    }
+
+    void selectionSort(Delivery* deliveryArray, const size_t arraySize) {
+        for (int i = 0; i < arraySize - 1; ++i) {
+            int minIndex = i;
+
+            for (int j = i + 1; j < arraySize; ++j) {
+
+                if (isSecondStringLarger(deliveryArray[j].address, deliveryArray[minIndex].address)) {
+                    minIndex = j;
+                }
+
+            }
+
+            if (minIndex != i) {
+                swapDeliveries(deliveryArray[i], deliveryArray[minIndex]);
+            }
+        }
+    }
+
+    void initializeDelivery(Delivery& delivery, const double price, const char* address) {
+        delivery.price = price;
+        assignAddress(delivery, address);
+    }
+
+    void writeDeliveryArrayToFile(std::ofstream& out, const Delivery* deliveryArray, const size_t arraySize) {
+        out.write((const char*)deliveryArray, sizeof(Delivery) * arraySize);
+    }
+
+    void writeSortedDeliveryArrayToFile(std::ofstream& out, Delivery* deliveryArray, const size_t arraySize) {
+        selectionSort(deliveryArray, arraySize);
+        out.write((const char*)deliveryArray, sizeof(Delivery) * arraySize);
```

Comment:
This function writes the array to a file, and the array is sorted as a side effect.
It is not a good practice to have 2 functionalities in 1 function. Better call selectionSort() in main() and then call writeDeliveryArrayToFile with the sorted array.